<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summary-based multivariate QTL modeling • zqtl</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Summary-based multivariate QTL modeling">
<meta property="og:description" content="Implementation of zQTL methods">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">zqtl</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><div id="summary-based-qtl-mapping" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#summary-based-qtl-mapping" class="anchor"></a>Summary-based QTL mapping</h1></div>
<div id="notes-on-mac-os-x-users" class="section level2">
<h2 class="hasAnchor">
<a href="#notes-on-mac-os-x-users" class="anchor"></a>Notes on Mac OS X users</h2>
<p>For some unknown reason, compiled Eigen library does not work as expected with <code>clang</code> compiler, e.g., producing lots of <code>nan</code> results. We recommend using <code>gcc</code> after version 5 that fully support <code>std=c++14</code> flag.</p>
</div>
<div id="required-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#required-packages" class="anchor"></a>Required packages</h2>
<p><code>Rcpp</code>, <code>RcppEigen</code>, <code>RcppProgress</code>, <code>RcppZiggurat</code>, <code>dplyr</code>, <code>tidyr</code></p>
</div>
<div id="installation-a-simple-version" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-a-simple-version" class="anchor"></a>Installation (a simple version)</h2>
<p>Just do this in <code>R</code></p>
<pre><code>&gt; install.packages('zqtl_x.x.x.tar.gz')</code></pre>
</div>
<div id="installation-optional" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-optional" class="anchor"></a>Installation (optional)</h2>
<p>Make sure your R development environment support <code>C++14</code> by including <code>-std=c++14</code> to <code>CFLAGS</code> and <code>CXXFLAGS</code> in <code>~/.R/Makevars</code> file. For instance,</p>
<pre><code>CXX = g++-6
CXXFLAGS = -O3 -std=c++14
CFLAGS = -O3 -std=c++14</code></pre>
<p>Build package locally.</p>
<pre><code>$ R CMD build .</code></pre>
<p>You will have <code>zqtl_x.x.x.tar.gz</code> gzipped file in current directory (<code>x.x.x</code> can be any version of the package). Install package within R:</p>
<pre><code>&gt; install.packages('zqtl_x.x.x.tar.gz')</code></pre>
<p>To speed up matrix-vector multiplication, one can compile codes with Intel MKL library. Currently <code>RcppEigen</code> does not support <code>BLAS</code> only options (this may not be true).</p>
<p>Enjoy!</p>
</div>
</div>
<div id="release-notes" class="section level1">
<h1 class="hasAnchor">
<a href="#release-notes" class="anchor"></a>Release notes</h1>
<div id="v-1-4-0" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-4-0" class="anchor"></a>v 1.4.0</h3>
<ul>
<li>Add useful utilities</li>
</ul>
</div>
<div id="v-1-3-6" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-3-6" class="anchor"></a>v.1.3.6</h3>
<ul>
<li>Improve variance calculation</li>
<li>Partitioned variance calculation with annotation information</li>
</ul>
</div>
<div id="v-1-3-5" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-3-5" class="anchor"></a>v.1.3.5</h3>
<ul>
<li>Report both types of residual variance estimation</li>
<li>Lower the precision limit in SGVB steps</li>
<li>Add adjusting routines for convenience</li>
</ul>
</div>
<div id="v-1-3-4" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-3-4" class="anchor"></a>v.1.3.4</h3>
<ul>
<li>Remove “backfire control” and “two step optimization”</li>
<li>Revive variance calculation with residual estimation</li>
<li>Make a room for univariate confounding factors</li>
<li>Output “clean” version of GWAS effects</li>
</ul>
</div>
<div id="v-1-3-3" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-3-3" class="anchor"></a>v.1.3.3</h3>
<ul>
<li>Minor fix on smoothness of the unmediated effect</li>
<li>Add <code>do.control.backfire</code> to regress out <code>M0 -&gt; M</code>.</li>
<li>Add experimental factorization model <code>factorization.model = 1</code>
</li>
</ul>
</div>
<div id="v-1-3-2" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-3-2" class="anchor"></a>v.1.3.2</h3>
<ul>
<li>Estimate unmediated factors by factorization</li>
<li>Boost sample size by resampling eigen vectors</li>
<li>Allow two step optimization</li>
<li>Parametric bootstrap for sensitivity analysis</li>
</ul>
</div>
<div id="v-1-3-1" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-3-1" class="anchor"></a>v.1.3.1</h3>
<ul>
<li>Make propensity sampling as second option</li>
<li>Estimate the unmediated effect as before</li>
<li>Drop weight features (not so useful)</li>
<li>Allow multiple mediators in the conditional analysis</li>
</ul>
</div>
<div id="v-1-3-0" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-3-0" class="anchor"></a>v.1.3.0</h3>
<ul>
<li>Counterfactual estimation of average unmediated effects via sampling</li>
<li>Factored mediation model</li>
</ul>
</div>
<div id="v-1-2-3" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-2-3" class="anchor"></a>v.1.2.3</h3>
<ul>
<li>Take multivariate effect sizes for mediator QTLs</li>
<li>Initialization by dot product</li>
</ul>
</div>
<div id="v-1-2-2" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-2-2" class="anchor"></a>v.1.2.2</h3>
<ul>
<li>Restrict number of mediator variables during the estimation of unmediated effects</li>
</ul>
</div>
<div id="v-1-2-1" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-2-1" class="anchor"></a>v.1.2.1</h3>
<ul>
<li>Estimation of average unmediated effects</li>
<li>Minor bugfix in regression (twice rescaling)</li>
</ul>
</div>
<div id="v-1-2-0" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-2-0" class="anchor"></a>v.1.2.0</h3>
<ul>
<li>Matrix factorization for confounder correction</li>
<li>Confirmed usefulness of non-negative parameters</li>
<li>Simplified pleiotropy model in mediation analysis</li>
<li>Adjust scales by standard deviation</li>
<li>Variance model in mediation analysis</li>
</ul>
</div>
<div id="v-1-1-0" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-1-0" class="anchor"></a>v.1.1.0</h3>
<ul>
<li>Spiked Gamma for the factored zQTL methods</li>
<li>Additional covariate component for confounder correction</li>
</ul>
</div>
<div id="v-1-0-1" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-0-1" class="anchor"></a>v.1.0.1</h3>
<ul>
<li>Variance calculation. See <code>note/variance.md</code> for details.</li>
<li>Random effect component to account for uncertainty of individuals</li>
<li>Removing LD-structure bias between mediation QTL and GWAS cohorts</li>
</ul>
</div>
<div id="v-1-0-0" class="section level3">
<h3 class="hasAnchor">
<a href="#v-1-0-0" class="anchor"></a>v.1.0.0</h3>
<ul>
<li>Initial version migrated from MIT github</li>
</ul>
</div>
</div>
<div id="bug-reports" class="section level1">
<h1 class="hasAnchor">
<a href="#bug-reports" class="anchor"></a>Bug reports</h1>
<p>Yongjin Park <code>ypp@csail.mit.edu</code></p>
</div>
<div id="technical-details" class="section level1">
<h1 class="hasAnchor">
<a href="#technical-details" class="anchor"></a>Technical details</h1>
<div id="standardization-of-summary-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#standardization-of-summary-statistics" class="anchor"></a>Standardization of summary statistics</h2>
<p>As we can read from Finucane <em>et al.</em> summary-based methods are quite sensitive to the prevalence of SNPs with unusually strong associations as they do not follow theoretical null distributions.</p>
<blockquote>
<p><strong>Outlier removal.</strong> To minimize standard error, we remove outlier SNPs by excluding SNPs with χ2 &gt; max{80, 0.001N}, where N is the maximum sample size in the study. We also remove the MHC region from all analyses, because of its unusual LD patterns and genetic architecture.</p>
</blockquote>
<p>For a univariate mediation analysis, the scale and meanshift of summary statistics do not substantially affect downstream analysis. However, for our model based analysis, we need z-scores of GWAS and mediation QTLs should behave similarly, just as a design matrix in Lasso regression is standardized.</p>
<p>Let’s consider a generative scheme with a single bias term <span class="math inline">\(\mu\)</span> and scaling by <span class="math inline">\(\tau\)</span>:</p>
<p><span class="math display">\[\mathbf{z} \sim \mathcal{N}\!\left(R (\mu I), \tau R\right)\]</span></p>
<p>where the SNP-SNP covariance matrix <span class="math inline">\(R = V^{\top} D^{2} V\)</span> was calculated from singular value decomposition of genotype matrix <span class="math inline">\(X/\sqrt{n} = UDV^{\top}\)</span>. An equivalent isotropic model could be derived after multiplying by <span class="math inline">\(D^{-1}V^{\top}\)</span>:</p>
<p><span class="math display">\[D^{-1} V^{\top}\mathbf{z} \sim \mathcal{N}\!\left(D V^{\top} I \mu, \tau I\right)\]</span></p>
<p>Letting <span class="math inline">\(\mathbf{y}\equiv D^{-1} V^{\top}\mathbf{z}\)</span> and <span class="math inline">\(\mathbf{x} \equiv D V^{\top} I\)</span>, we can find the maximum likelihood estimate <span class="math inline">\(\hat{\mu}\)</span> by minimizing</p>
<p><span class="math display">\[- \frac{1}{2} \ln \tau -\frac{1}{2\tau} \sum_{k} \left(y_{k} - x_{k} \mu\right)^{2},\]</span></p>
<p>which leads to</p>
<p><span class="math display">\[\hat{\mu} = \frac{\sum_{k} y_{k} x_{k}}{\sum_{k} x_{k}^{2}} = \frac{\mathbf{x}^{\top}\mathbf{y}}{\mathbf{x}^{\top}\mathbf{x}} = \frac{\mathbf{z}^{\top}I}{I^{\top}RI}\]</span></p>
<p>and</p>
<p><span class="math display">\[\hat{\tau} = \frac{1}{K - 1}\|\mathbf{y} - \hat{\mu}\mathbf{x}\|^{2}\]</span></p>
<p>Using these, we standardize</p>
<p><span class="math display">\[\mathbf{z} \gets (\mathbf{z} - \hat{\mu} RI)/\sqrt{\hat{\tau}}\]</span></p>
</div>
<div id="simulation" class="section level2">
<h2 class="hasAnchor">
<a href="#simulation" class="anchor"></a>Simulation</h2>
<p>Let’s take a look at Kichaev <em>et al.</em> (2014)</p>
<blockquote>
<p>Once we established the causal SNPs, we used a linear model to simulate continuous phenotypes such that the causal SNPs aggregated to explain a fixed proportion of the phenotypic variance (<span class="math inline">\(h_{g}^{2}\)</span>). This phenotypic variance was partitioned equally amongst all the causal SNPs (qualitatively similar results were obtained when phenotypic variance was unevenly partitioned among causal variants (see Figure S6)). In particular, the individual’s phenotype was drawn according to <span class="math inline">\(Y_{m} = \sum_{i=1}^{N_{c}}\beta_{i} G_{im} + \epsilon_{m}\)</span>, where <span class="math inline">\(N_{c}\)</span> is the total number of causal variants, <span class="math inline">\(\beta_{i}\)</span> is the effect size of the ith causal SNP, <span class="math inline">\(G_{im}\)</span> is number of copies of the risk allele <span class="math inline">\(i\)</span> (randomly assigned as reference or alternate) for individual m, and <span class="math inline">\(\epsilon_{m} \sim \mathcal{N}\!\left(0, 1 - h_{g}^{2}\right)\)</span>. Finally, we calculated association Z-scores (<span class="math inline">\(Z_{ij}\)</span>) at each SNP <span class="math inline">\(i,j\)</span> by taking the Wald statistic from the regression of the Y on <span class="math inline">\(G_{ij}\)</span>, where Y is a vector of phenotypes for M individuals and <span class="math inline">\(G_{ij}\)</span> is the vector of corresponding genotypes for the ith SNP at the jth locus. For simulations that required loci greater than 10 KB, we instead drew Z-scores from a Multivariate Normal distribution with covariance equal to LD based on the European 1 KG and non-centrality parameters at causal sites drawn from a Normal distribution with mean 5 and standard deviation 0.2.</p>
</blockquote>
<p>From Kichaev <em>et al.</em> (2017)</p>
<blockquote>
<p>We simulated phenotypes under a linear model such that for individual i of population p, their phenotype Y was drawn as <span class="math inline">\(Y_{ip} = \sum_{j=1}^{N_{c}} \beta_{j} G_{ijp} + \epsilon_{ip}\)</span>, where <span class="math inline">\(N_{c}\)</span> is the total number of causal variants, <span class="math inline">\(\beta_{j}\)</span> is the effect size of the jth causal SNP, <span class="math inline">\(G_{ijp}\)</span> is the number of copies of the risk allele j for individual i of population p. Following recent works, we simulated similar heritability across populations The population-specific error term, <span class="math inline">\(\epsilon_{ip}\)</span>, was drawn according to a <span class="math inline">\(\mathcal{N}\!\left(0,\sigma_{ep}^{2}\right)\)</span>, where <span class="math inline">\(\sigma_{ep}^{2} = (\sigma_{gp}^{2} - h_{g}^{2} \sigma_{gp}^{2})/h_{g}^{2}\)</span>, <span class="math inline">\(\sigma_{gp}^{2} = \boldsymbol{\beta}^{\top}\mathsf{Cov}(X_{p})\boldsymbol{\beta}\)</span>, and <span class="math inline">\(\mathsf{Cov}(X_{p})\)</span> is the population-specific covariance of the genotypes (LD). The effect size, <span class="math inline">\(\beta_{j}\)</span>, was set to be inversely proportional to the average SD of the population allele frequencies; this is roughly equivalent to assuming that each causal SNP explains an equal proportion of the phenotypic variance.</p>
</blockquote>
</div>
<div id="correction-of-non-genetic-correlations-between-z-scores" class="section level2">
<h2 class="hasAnchor">
<a href="#correction-of-non-genetic-correlations-between-z-scores" class="anchor"></a>Correction of non-genetic correlations between z-scores</h2>
<p>Suppose we have a phenotype vector generated by two components, genotype matrix <code>G</code> and unwanted confounder matrix <code>U</code>:</p>
<pre><code>y = G θ + U δ + G0 α</code></pre>
<p>Let’s take the advantages of this fact <code>G0 t(G) -&gt; 0</code>. Now consider two z-scores of the same number of SNPs:</p>
<pre><code>z ~ t(G) y = t(G)G θ + t(G) U δ + t(G) G0 α
           -&gt; R θ + t(G) U δ</code></pre>
<p>and</p>
<pre><code>z0 ~ t(G0) y = t(G0)G θ + t(G) U δ + t(G0) G0 α
           -&gt; λ I θ + t(G0) U δ + R0 α</code></pre>
<p>Therefore correlation between <code>z</code> and <code>z0</code> will approach to <code>δ^2||u||</code> up to some constant factor.</p>
</div>
<div id="removing-discrepancy-between-two-reference-panels-in-mediation-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#removing-discrepancy-between-two-reference-panels-in-mediation-analysis" class="anchor"></a>Removing discrepancy between two reference panels in mediation analysis</h2>
<p>Assume <code>X1 / sqrt(n1) = U1 D1 t(V1)</code> and <code>X / sqrt(n) = U D t(V)</code>.</p>
<pre><code>E[α.hat]           = S1 R1 inv(S1) α</code></pre>
<p>therefore</p>
<pre><code>E[α]               = S1 inv(R1) inv(S1) α.hat
                   = S1 V1 inv(D1^2) t(V1) inv(S1) α.hat</code></pre>
<p>Construct a “design matrix” for the mediation effect:</p>
<pre><code>t(V) R inv(S) E[α] = t(V) V D^2 t(V) inv(S) α
                   = D^2 t(V) inv(S) S1 V1 inv(D1^2) t(V1) inv(S1) α.hat
M                  = D^2 t(V) inv(S) S1 (V1/D1) t(inv(D1)V1) * Zα</code></pre>
<p>Mediation regression model assuming mean-field distribution of <code>β</code>:</p>
<pre><code>E[Δ] = M E[β]
V[Δ] = M^2 V[β]</code></pre>
<p>For a perfect mediation:</p>
<pre><code>E[t(V)Zθ] = E[Δ]</code></pre>
</div>
<div id="variance-calculation" class="section level2">
<h2 class="hasAnchor">
<a href="#variance-calculation" class="anchor"></a>Variance calculation</h2>
<div id="notations" class="section level3">
<h3 class="hasAnchor">
<a href="#notations" class="anchor"></a>Notations</h3>
<ul>
<li><p>p : the number of SNPs</p></li>
<li><p>n : the number of individuals</p></li>
<li><p>n0 : the number of individuals in the mediation QTL models</p></li>
<li><p>K : the number of mediators</p></li>
<li><p>X : a standardized genotype matrix (n x p); mean = 0 and variance 1</p></li>
<li><p>y : phenotype vector (n x 1)</p></li>
<li><p>z : z-score vector of XWAS (p x 1)</p></li>
<li><p>a : the estimated univariate effect size of mediation QTLs (p x K)</p></li>
<li><p>S0 : the diagonal matrix of standard errors (p x p)</p></li>
<li><p>za : the z-score matrix for mediation QTLs</p></li>
<li><p>α : multivariate effect size of mediation QTLs</p></li>
<li><p>β : mediation effect size (K x 1)</p></li>
<li><p>γ : unmediated effect size (p x 1)</p></li>
</ul>
</div>
<div id="estimation-of-residual-effects-on-reference-panel" class="section level3">
<h3 class="hasAnchor">
<a href="#estimation-of-residual-effects-on-reference-panel" class="anchor"></a>Estimation of residual effects on reference panel</h3>
<p>From the linear model <code>y = X θ + ε</code>, we can derive the estimated univariate statistics.</p>
<pre><code>θ.uni = t(X)y / n
      = t(X)(X θ + ε) / n
      = t(X)X θ / n + t(X)ε / n</code></pre>
<p>With <code>S = σ / sqrt(n)</code> and exploiting SVD <code>X/sqrt(n) = UD t(V)</code>,</p>
<pre><code>z     = θ.uni * sqrt(n) / σ
      = R (θ * sqrt(n)/σ) + t(X/sqrt(n)) ε /σ 

t(V)z = t(V) V D^2 t(V) (θ * sqrt(n)/σ) + t(V)VD t(U) (ε/σ)
      = D^2 t(V) (θ * sqrt(n)/σ) + D^2 inv(D)t(U) (ε/σ)</code></pre>
<p>Treating <code>y.tilde = t(V)z</code> and <code>Δ = inv(D)t(U)</code>, we can estimate the individual-level residual values <code>ε</code> fitting the regression :</p>
<pre><code>y.tilde ~ ... + D^2 Δ (ε/σ) + N(0, D^2)</code></pre>
<p>Then we can estimate the scaled residual variance by <code>V[ε/σ]</code>, and we can approximate that value with <code>sum(ε^2)/nσ^2</code> if <code>E[ε_i] = 0</code>. For unbiased reference panel, it is not so hard to see that <code>V[ε/σ] -&gt; 1</code>.</p>
</div>
</div>
<div id="variance-components" class="section level2">
<h2 class="hasAnchor">
<a href="#variance-components" class="anchor"></a>Variance components</h2>
<p>We define variance explained by the linear model on the reference cohort genotype matrix, <code>var_model = V[Xθ]/σ^2</code>. Defining <code>η = Xθ</code> we can estimate</p>
<pre><code>V[Xθ] = Σ η_j^2/n - (Σ η_j/n)^2.</code></pre>
<p>Alternatively we could use <code>t(θ) R θ</code>, but we may not have <code>Ε[Xθ] = 0</code> in practice. Variance unexplained:</p>
<pre><code>var_residual = V[ε/σ] ~ Σ ε_i^2 / nσ^2</code></pre>
<div id="reconciling-different-samples-sizes-between-the-mediation-and-gwas-models" class="section level3">
<h3 class="hasAnchor">
<a href="#reconciling-different-samples-sizes-between-the-mediation-and-gwas-models" class="anchor"></a>Reconciling different samples sizes between the mediation and GWAS models</h3>
<p>The model for mediated QTLs:</p>
<pre><code>E[a]  = S0 R inv(S0)α
α     = S0 inv(R) inv(S0) E[a]</code></pre>
<p>For simplicity we assume <code>S = (σ/sqrt(n))I</code> and <code>S0 = (σ0/sqrt(n0))I</code> and <code>X</code> is standardized.</p>
<pre><code>E[θ.uni] = S R inv(S) θ
         = S R inv(S) (αβ + γ)
         = S R inv(S) S0 inv(R) inv(S0) E[a] β + S R inv(S) γ
         = E[a] β + S R inv(S) γ
E[z]     = inv(S) E[a] β + R inv(S) γ
t(V)E[z] = (t(V) inv(S) E[a]) β + D^2 t(V) inv(S) γ
         = (t(V) Λ inv(S0) E[a]) β + D^2 t(V) inv(S) γ
where
Λ        = sqrt(n)/sqrt(n0) I</code></pre>
<ul>
<li>Variance explained by mediated component:</li>
</ul>
<pre><code>αβ            = inv(R) E[a] β = V inv(D^2) t(V) E[a] β
var_mediation = Σ (αβ)_j^2 / σ^2</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> License</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Yongjin Park <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Yongjin Park.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
