# Technical details

## Standardization of summary statistics

As we can read from Finucane _et al._ summary-based methods are quite
sensitive to the prevalence of SNPs with unusually strong associations
as they do not follow theoretical null distributions.

> __Outlier removal.__ To minimize standard error, we remove outlier
> SNPs by excluding SNPs with χ2 > max{80, 0.001N}, where N is the
> maximum sample size in the study. We also remove the MHC region from
> all analyses, because of its unusual LD patterns and genetic
> architecture.

For a univariate mediation analysis, the scale and meanshift of
summary statistics do not substantially affect downstream analysis.
However, for our model based analysis, we need z-scores of GWAS and
mediation QTLs should behave similarly, just as a design matrix in
Lasso regression is standardized.

Let's consider a generative scheme with a single bias term $\mu$ and
scaling by $\tau$:

$$\mathbf{z} \sim \mathcal{N}\!\left(R (\mu I), \tau R\right)$$

where the SNP-SNP covariance matrix $R = V^{\top} D^{2} V$ was
calculated from singular value decomposition of genotype matrix
$X/\sqrt{n} = UDV^{\top}$.  An equivalent isotropic model could be
derived after multiplying by $D^{-1}V^{\top}$:

$$D^{-1} V^{\top}\mathbf{z} \sim \mathcal{N}\!\left(D V^{\top} I \mu, \tau I\right)$$

Letting $\mathbf{y}\equiv D^{-1} V^{\top}\mathbf{z}$ and $\mathbf{x} \equiv D V^{\top} I$,
we can find the maximum likelihood estimate $\hat{\mu}$ by minimizing

$$- \frac{1}{2} \ln \tau -\frac{1}{2\tau} \sum_{k} \left(y_{k} - x_{k} \mu\right)^{2},$$

which leads to 

$$\hat{\mu} = \frac{\sum_{k} y_{k} x_{k}}{\sum_{k} x_{k}^{2}} = \frac{\mathbf{x}^{\top}\mathbf{y}}{\mathbf{x}^{\top}\mathbf{x}} = \frac{\mathbf{z}^{\top}I}{I^{\top}RI}$$

and

$$\hat{\tau} = \frac{1}{K - 1}\|\mathbf{y} - \hat{\mu}\mathbf{x}\|^{2}$$

Using these, we standardize

$$\mathbf{z} \gets (\mathbf{z} - \hat{\mu} RI)/\sqrt{\hat{\tau}}$$


## Simulation

Let's take a look at Kichaev _et al._ (2014)

> Once we established the causal SNPs, we used a linear model to
> simulate continuous phenotypes such that the causal SNPs aggregated
> to explain a fixed proportion of the phenotypic variance
> ($h_{g}^{2}$). This phenotypic variance was partitioned equally
> amongst all the causal SNPs (qualitatively similar results were
> obtained when phenotypic variance was unevenly partitioned among
> causal variants (see Figure S6)). In particular, the individual's
> phenotype was drawn according to $Y_{m} =
> \sum_{i=1}^{N_{c}}\beta_{i} G_{im} + \epsilon_{m}$, where $N_{c}$ is
> the total number of causal variants, $\beta_{i}$ is the effect size
> of the ith causal SNP, $G_{im}$ is number of copies of the risk
> allele $i$ (randomly assigned as reference or alternate) for
> individual m, and $\epsilon_{m} \sim \mathcal{N}\!\left(0, 1 -
> h_{g}^{2}\right)$. Finally, we calculated association Z-scores
> ($Z_{ij}$) at each SNP $i,j$ by taking the Wald statistic from the
> regression of the Y on $G_{ij}$, where Y is a vector of phenotypes
> for M individuals and $G_{ij}$ is the vector of corresponding
> genotypes for the ith SNP at the jth locus. For simulations that
> required loci greater than 10 KB, we instead drew Z-scores from a
> Multivariate Normal distribution with covariance equal to LD based
> on the European 1 KG and non-centrality parameters at causal sites
> drawn from a Normal distribution with mean 5 and standard deviation
> 0.2.

From Kichaev _et al._ (2017)

> We simulated phenotypes under a linear model such that for
> individual i of population p, their phenotype Y was drawn as $Y_{ip}
> = \sum_{j=1}^{N_{c}} \beta_{j} G_{ijp} + \epsilon_{ip}$, where
> $N_{c}$ is the total number of causal variants, $\beta_{j}$ is the
> effect size of the jth causal SNP, $G_{ijp}$ is the number of copies
> of the risk allele j for individual i of population p. Following
> recent works, we simulated similar heritability across populations
> The population-specific error term, $\epsilon_{ip}$, was drawn
> according to a $\mathcal{N}\!\left(0,\sigma_{ep}^{2}\right)$, where
> $\sigma_{ep}^{2} = (\sigma_{gp}^{2} - h_{g}^{2}
> \sigma_{gp}^{2})/h_{g}^{2}$, $\sigma_{gp}^{2} =
> \boldsymbol{\beta}^{\top}\mathsf{Cov}(X_{p})\boldsymbol{\beta}$, and
> $\mathsf{Cov}(X_{p})$ is the population-specific covariance of the
> genotypes (LD). The effect size, $\beta_{j}$, was set to be
> inversely proportional to the average SD of the population allele
> frequencies; this is roughly equivalent to assuming that each causal
> SNP explains an equal proportion of the phenotypic variance.


## Correction of non-genetic correlations between z-scores

Suppose we have a phenotype vector generated by two components,
genotype matrix `G` and unwanted confounder matrix `U`:

```
y = G θ + U δ + G0 α
```

Let's take the advantages of this fact `G0 t(G) -> 0`.  Now consider
two z-scores of the same number of SNPs:

```
z ~ t(G) y = t(G)G θ + t(G) U δ + t(G) G0 α
           -> R θ + t(G) U δ
```
and

```
z0 ~ t(G0) y = t(G0)G θ + t(G) U δ + t(G0) G0 α
           -> λ I θ + t(G0) U δ + R0 α
```

Therefore correlation between `z` and `z0` will approach to `δ^2||u||` up
to some constant factor.


## Removing discrepancy between two reference panels in mediation analysis

Assume `X1 / sqrt(n1) = U1 D1 t(V1)` and `X / sqrt(n) = U D t(V)`.

```
E[α.hat]           = S1 R1 inv(S1) α
```
therefore
```
E[α]               = S1 inv(R1) inv(S1) α.hat
                   = S1 V1 inv(D1^2) t(V1) inv(S1) α.hat
```
Construct a "design matrix" for the mediation effect:
```
t(V) R inv(S) E[α] = t(V) V D^2 t(V) inv(S) α
                   = D^2 t(V) inv(S) S1 V1 inv(D1^2) t(V1) inv(S1) α.hat
M                  = D^2 t(V) inv(S) S1 (V1/D1) t(inv(D1)V1) * Zα
```
Mediation regression model assuming mean-field distribution of `β`:
```
E[Δ] = M E[β]
V[Δ] = M^2 V[β]
```
For a perfect mediation:
```
E[t(V)Zθ] = E[Δ]
```

